robot-server/robot_server/runs/run_store.py

```diff
@@ -537,6 +551,106 @@ def get_all_commands_as_preserialized_list(
+        """Get run commands errors count from the store.
+
+        Args:
+            run_id: Run ID to pull commands from.
+
+        Returns:
+            The number of commands errors.
+
+        Raises:
+            RunNotFoundError: The given run ID was not found.
+        """
+        with self._sql_engine.begin() as transaction:
+            if not self._run_exists(run_id, transaction):
+                raise RunNotFoundError(run_id=run_id)
+
+            select_count = sqlalchemy.select(sqlalchemy.func.count()).where(
+                and_(
+                    run_command_table.c.run_id == run_id,
+                    run_command_table.c.command_status == CommandStatusSQLEnum.FAILED,
+                )
+            )
+            errors_count: int = transaction.execute(select_count).scalar_one()
+            return errors_count
+
+    def get_commands_errors_slice(
+        self,
+        run_id: str,
+        length: int,
+        cursor: Optional[int],
+    ) -> CommandErrorSlice:
+        """Get a slice of run commands errors from the store.
+
+        Args:
+            run_id: Run ID to pull commands from.
+            length: Number of commands to return.
+            cursor: The starting index of the slice in the whole collection.
+                If `None`, up to `length` elements at the end of the collection will
+                be returned.
+
+        Returns:
+            A collection of command errors as well as the actual cursor used and
+            the total length of the collection.
+
+        Raises:
+            RunNotFoundError: The given run ID was not found.
+        """
+        with self._sql_engine.begin() as transaction:
+            if not self._run_exists(run_id, transaction):
+                raise RunNotFoundError(run_id=run_id)
+
+            select_count = sqlalchemy.select(sqlalchemy.func.count()).where(
+                and_(
+                    run_command_table.c.run_id == run_id,
+                    run_command_table.c.command_status == CommandStatusSQLEnum.FAILED,
+                )
+            )
+            count_result: int = transaction.execute(select_count).scalar_one()
+
+            actual_cursor = cursor if cursor is not None else count_result - length
+            # Clamp to [0, count_result).
+            # cursor is 0 based index and row number starts from 1.
+            actual_cursor = max(0, min(actual_cursor, count_result - 1)) + 1
+            select_command_errors = (
+                sqlalchemy.select(
+                    sqlalchemy.func.row_number().over().label("row_num"),
+                    run_command_table,
+                )
+                .where(
+                    and_(
+                        run_command_table.c.run_id == run_id,
+                        run_command_table.c.command_status
+                        == CommandStatusSQLEnum.FAILED,
+                    )
+                )
+                .subquery()
```

Comment:
I think this subquery needs an `ORDER BY` clause too, right? Because you're depending on the incrementing numbers from `sqlalchemy.func.row_number()` corresponding to the actual order of the commands.

```suggestion
                .order_by(run_command_table.c.index_in_run)
                .subquery()
```